### ISSUE TITLE
Not null invariant incorrectly fails on non-nullable field inside a nullable struct

### ISSUE THREAD

**Issue by Kimahriman** [delta-io/delta#860]
An existing test actually tests for what I think is incorrect behavior: https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/schema/InvariantEnforcementSuite.scala#L186

```
 testQuietly("reject non-nullable nested column") {
    val schema = new StructType()
      .add("top", new StructType()
        .add("key", StringType, nullable = false)
        .add("value", IntegerType))
    testBatchWriteRejection(
      NotNull(Seq("key")),
      schema,
      spark.createDataFrame(Seq(Row(Row("a", 1)), Row(Row(null, 2))).asJava, schema.asNullable),
      "top.key"
    )
    testBatchWriteRejection(
      NotNull(Seq("key")),
      schema,
      spark.createDataFrame(Seq(Row(Row("a", 1)), Row(null)).asJava, schema.asNullable),
      "top.key"
    )
  }
```

Here the `top` struct is nullable but the `key` field is not, and the current invariant checks only care about the fact that `key` is non-nullable therefore selecting that value (through a series of `GetStructField`'s), will always not be null. However, it is valid for `top` to be null, and it's more accurate to say that `key` is never null when `top` is not null I think.

So in this test case, the first one is a valid test case. `top` is not null but `key` is. However, in the second test case, `top` is null, which should be valid behavior and not throw an exception I believe.

After looking through the code I can see a few ways to make it basically skip checking `key` in this case, but it might be more ideal but more complicated to have it check `top` first, and only if that's not null, then check `key` and fail only in that case.

**Comment by scottsand-db** [comment]
Hi @Kimahriman, thanks for bringing this to our attention. We will look into this.

**Comment by brkyvz** [comment]
I believe this is the behavior that most databases follow. The idea was that if `foo.bar` is not null, then `foo` should always be not null. If you want the case that `foo` can be null, but whenever foo is not null, then bar should also be not null, then you should be using a check constraint such as:
```
foo is null or foo.bar is not null
```

**Comment by Kimahriman** [comment]
Spark seems to allow for that behavior of foo being nullable but foo.bar not

**Comment by Kimahriman** [comment]
Yeah I'm just saying this _is_ valid Spark behavior, so you can hit an exception for valid Spark code. Simplest reproduction:

```
import pyspark.sql.functions as F
from delta.tables import DeltaTables
test = spark.range(1).withColumn('nested', F.when(F.col('id') < 5, F.struct('id')))
test.printSchema()

root
|-- id: long (nullable = false)
|-- nested: struct (nullable = true)
|    |-- id: long(nullable = false)

DeltaTable.createOrReplace().location('file:///tmp/test').addColumns(test.schema).execute()
test.write.format('delta').mode('append').save('file:///tmp/test')
```

**Comment by Kimahriman** [comment]
And Spark correctly updates the nullability when it needs to:

```
test.select('nested.*').printSchema()

root
|-- id: long (nullable = true)
```

**Comment by Kimahriman** [comment]
> if users set a not-null constraint, they likely will expect they should never see a null value when writing the code

That's the main issue, not-null constraints are automatically applied with no way to disable them. I have been slowly playing around with ways update the not-null constraint checking to operate in-line with how Spark works, haven't made too much progress on it though. At the very least it should be an opt-in thing to get around this. I still consider it a bug :)

**Comment by zsxwing** [comment]
We prefer to avoid creating non standard behaviors. Could you clarify what use cases that may require this non standard behavior?

**Comment by zsxwing** [comment]
> it always has a value iff the string input is not null

Can you set the nested field nullable and add a check constraint such as `nested is null or nested.id is not null` to the table instead?

**Comment by zsxwing** [comment]
Adding a new behavior even if it's opt-in would also add extra work when we expend the not null constraint behaviors to other engines. And I feel it's not worth right now.

I will keep this open to hear more voice.

**Comment by megri** [comment]
I was bitten by this today. I'm using Scala and have this model

```scala
case class MyRow(id: Int, value: String)

case class Changes(before: Option[MyRow], after: MyRow)
```

modelling changes between `MyRow`s of the same id. I expected this to work but get `org.apache.spark.sql.streaming.StreamingQueryException: NOT NULL constraint violated for column: before.id.`

**Comment by zsxwing** [comment]
Yep. It would be great if anyone familiar with Presto/Trino can provide the feedback.

**Comment by Kimahriman** [comment]
My team continues to get bit by this repeatedly, and we have to find workarounds to force certain fields to be nullable even when they don't have to be, and I see more people bring this up (someone just did on Slack). 

Any reconsideration on respecting struct nullability? I still don't see how there's any valid way to consider that a non nullable field inside a nullable struct has to always be not null. That's an invalid interpretation of the struct being nullable. And it isn't even consistent with how Parquet files treat nullable structs. According to the user on slack at least Kafka interprets this correctly

**Comment by PhillHenry** [comment]
> Yeah there are workarounds, just less than ideal. You lose the 0.01% speedup you would get by avoiding null-checks on the field hah. 

You'd be surprised. Typically, code will check a single pointer for null and that pointer might have been assigned to something that lives in the CPU's cache (ie, access to it will be super fast). Checking for null with big data generally pulls large amounts from RAM (relatively slow) and pushes anything out of the CPU cache. If you do this for a large amount of data, it starts to show in in terms of performance.

**Comment by jqmtor** [comment]
This is a problem for my team as well and I think [this example](https://github.com/delta-io/delta/issues/860#issuecomment-1297006499) demonstrates it well. 

We are using Delta to transport data in Debezium CDC payloads. In that format, the top-level `before` and `after` fields need to be nullable. A delete event is denoted by a `null` after field and a create event is denoted by a `null` before field.

However, both those fields follow the same schema specification and, more importantly, the schema of the `before` and `after` fields is independent from the schema of the envelope that encloses them. We define the expected schema of the envelope by composing the schema of the entity (the `before` and `after` fields), with the schema of the envelope (the `op`, `source`, etc. fields).
This feature effectively ties the schema of the entity with the schema of the enclosing envelope. For our use cases, this coupling is problematic in several ways. The schema of the entities transported in the most important thing and it is used by several systems. Preserving their schema when transported via a Debezium CDC payload allows us to preserve their integrity and is the most important thing. These schemas are used in Delta, but also by other systems.

**Comment by Kimahriman** [comment]
Ironically that's exactly how the Delta transaction log is stored as well. `add` will only be non-null if it's an add action, `remove` will only be non-null if it's a `remove` action, etc.

Based on the way Delta handles nullability currently, there is no way to actually express the equivalent of "if the action is `add`, then `add.path` must be non-null". Trying to do that makes Delta interpret this as `add.path` must be non null even if the action is not `add`.

**Comment by Kimahriman** [comment]
Also FWIW this isn't an issue in delta-rs (they don't do any implicit null checking AFAIK). So you can write this "valid" data in other ways and then read in Spark, you just can't write it in Spark right now.

**Comment by foss-contributor** [comment]
@Kimahriman, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

**Comment by Kimahriman** [comment]
This is still relevant
